contracts/implementation/PoolKeeper.sol

```diff
@@ -162,6 +170,30 @@ contract PoolKeeper is IPoolKeeper, Ownable {
         }
     }
 
+    /**
+     * @notice Called by keepers to perform an update on multiple pools
+     * @param pools A tightly packed bytes array of LeveragedPool addresses to be upkept
+     *     20 bytes       20 bytes       20 bytes     ...
+     * | pool address | pool address | pool address | ... |
+     * @dev Arguments can be encoded with `L2Encoder.encodePerformUpkeepParams`
+     * @dev Will revert if the bytes array is a correct length (some multiple of 20 bytes)
+     */
+    function performUpkeepMultiplePoolsPacked(bytes calldata pools) external override {
+        require(pools.length % ADDRESS_LENGTH == 0, "Data must only include addresses");
+        uint256 poolsLength = pools.length / ADDRESS_LENGTH;
```

Comment:
```suggestion
        uint256 numPools = pools.length / ADDRESS_LENGTH;
```

I get the point but I think `poolsLength` is misleading as it's not really the length of `pools` but the number of actual pool addresses specified in `pools`.

More broadly, I think we should consider more opaque naming for the actual encoded parameters (as they are, after all, opaque data). For instance,

```solidity
function performUpkeepMultiplePoolsPacked(bytes calldata data) external override {
    require(data.length % ADDRESS_LENGTH == 0, "Data misaligned");

    uint256 numPools = data.length / ADDRESS_LENGTH;

    for (uint256 i = 0; i < numPools; i++) {
        performUpkeepSinglePool(getNextPool(ARRAY_START_OFFSET + i * ADDRESS_LENGTH));
    }
}
```

This makes it much clearer that we're (initially) referring to the actual length of the parameter data *in bytes* rather than the number of (decoded) parameters therein.

Reply:
I'm not sure I like the new error message. We are giving opaque data to the function, but what we are expecting is very clear so having a clear error message here seems like the best way to go IMO.